app/apis/api_admin.py

```diff
@@ -0,0 +1,139 @@
+        if role is None:
+            # applications which don't have users are also accesssible as admin
+            roles = db.session.query(ApplicationUserRole).filter(
+                ApplicationUserRole.application_id == application_id).count()
+            if roles == 0:
+                pass
+            else:
+                raise ApplicationUserRoleException
+        elif role.role != Role.admin:
+            raise ApplicationUserRoleException
+        return fn(*args, **kwargs)
+    return wrapper
+
+
+@admin_info_namespace.route('/<int:application_id>/acl')
+class ApiApplicationIdACL(Resource):
+    method_decorators = [check_role]
+
+    save_acl_parser = reqparse.RequestParser()
+    save_acl_parser.add_argument('uid', type=str, required=True, location='form')
+    save_acl_parser.add_argument('role', type=str, required=True, location='form')
+
+    delete_acl_parser = reqparse.RequestParser()
+    delete_acl_parser.add_argument('uid', type=str, required=True, location='form')
+
+    @admin_info_namespace.marshal_list_with(role_info)
+    def get(self, application_id):
+        roles = ApplicationUserRole.query.filter_by(application_id=application_id).all()
+        return roles
+
+    @admin_info_namespace.marshal_with(success_or_not)
+    @admin_info_namespace.expect(save_acl_parser)
+    def post(self, application_id):
+        args = self.save_acl_parser.parse_args()
+        uid = args['uid']
+        uobj = db.session.query(User).filter(User.user_uid == uid).one_or_none()
+        if uobj is None:
+            return {"status": False}, 400
+
+        roleObj = ApplicationUserRole(
+            application_id=application_id,
+            user_id=uobj.user_id,
+            role=args['role'])
+        db.session.add(roleObj)
+        db.session.flush()
+
+        # if role is added by first user, they will be admin
+        roles = db.session.query(ApplicationUserRole).filter(
+            ApplicationUserRole.application_id == application_id).count()
+        if roles <= 1:
+            sender_id = get_jwt_identity()
+            if uobj.user_id != sender_id:
+                adminObj = ApplicationUserRole(
+                    application_id=application_id,
+                    user_id=sender_id,
+                    role='admin')
+                db.session.add(adminObj)
+                db.session.flush()
+            else:
+                roleObj.role = 'admin'
+
+        db.session.commit()
+        db.session.close()
+        return {"status": True, "message": "Success."}, 200
+
+    @admin_info_namespace.marshal_with(success_or_not)
+    @admin_info_namespace.expect(save_acl_parser)
+    def patch(self, application_id):
+        args = self.save_acl_parser.parse_args()
+        uid = args['uid']
+        uobj = db.session.query(User).filter(User.user_uid == uid).one_or_none()
+        if uobj is None:
+            return {"status": False}, 400
+        roleObj = db.session.query(ApplicationUserRole).filter(
```

Comment:
We need to check the user' role before this line.

Reply:
ditto